crates/agentgateway/src/cel/tests.rs

```diff
@@ -209,3 +351,114 @@ fn test_properties() {
 	test(r#"extauthz.user_id"#, &["extauthz.user_id"]);
 	test(r#"extauthz.role == "admin""#, &["extauthz.role"]);
 }
+
+// ============================================================================
+// Comprehensive Benchmarks
+// ============================================================================
+
+// Benchmark: Compile phase - Expression::new() for each test case
+#[divan::bench]
+fn bench_comprehensive_compile(b: Bencher) {
+	let cases = test_cases();
+	b.bench(|| {
+		for (expr_str, _, _) in &cases {
+			let _ = divan::black_box(Expression::new(expr_str).unwrap());
+		}
+	});
+}
+
+// Benchmark: Build phase - ContextBuilder::build() for each test case
+#[divan::bench]
+fn bench_comprehensive_build(b: Bencher) {
+	let cases = test_cases();
+	// Pre-compile expressions
+	let compiled: Vec<_> = cases
+		.iter()
+		.map(|(expr_str, req_builder, _)| {
+			let expr = Expression::new(expr_str).unwrap();
+			let req = req_builder();
+			(expr, req)
+		})
+		.collect();
+
+	b.bench(|| {
+		for (expr, req) in &compiled {
+			let mut cb = ContextBuilder::new();
+			cb.register_expression(expr);
+			cb.with_request(req, "".to_string());
+			let _ = divan::black_box(cb.build().unwrap());
+		}
+	});
+}
+
+// Benchmark: Execute phase - exec.eval() for each test case
+#[divan::bench]
+fn bench_comprehensive_execute(b: Bencher) {
+	let cases = test_cases();
+	// Pre-compile and build contexts
+	let prepared: Vec<_> = cases
+		.iter()
+		.map(|(expr_str, req_builder, _)| {
+			let expr = Expression::new(expr_str).unwrap();
+			let req = req_builder();
+			let mut cb = ContextBuilder::new();
+			cb.register_expression(&expr);
+			cb.with_request(&req, "".to_string());
+			let exec = cb.build().unwrap();
+			(expr, exec)
+		})
+		.collect();
+
+	b.bench(|| {
+		for (expr, exec) in &prepared {
+			let _ = divan::black_box(exec.eval(expr).unwrap());
+		}
+	});
+}
+
+// Individual benchmarks for each phase on a single representative expression
+#[divan::bench]
+fn bench_single_compile(b: Bencher) {
+	let expr_str = r#"request.method == "GET" && request.headers["x-custom"] == "value""#;
+	b.bench(|| {
```

Comment:
hmm I actually want one subtest for each using divan::bench(args = cases).

Make cases store the vec of names, and allow looking up the cases by name (but keep in a vec for ordering)